### PR TITLE
Issue264 make kpi cal efficient

### PR DESCRIPTION
**Pull request by JavierArroyoBastida**
This is for https://github.com/ibpsa/project1-boptest/issues/264: to reduce the memory usage related to the calculation of KPIs, following what has been discussed in https://github.com/ibpsa/project1-boptest/issues/240. Specifically:

1. It allows to call the KPI Calculator iteratively, which was technically possible before but computational time was increasing with simulation time as integration always happened since the starting simulation time. Now, storage variables are included so that new integrations can be calculated from the results of previous integrations. This is particularly useful for the implementation of reinforcement learning algorithms (see https://github.com/ibpsa/project1-boptest-gym), where KPIs may need to be calculated every sampling time to define a reward function. Notice that in https://github.com/ibpsa/project1-boptest/issues/240 we were considering two possibilities: the integration within Modelica models and the use of cumulative records. I've finally gone for the latter because this one does not intrude into the emulator models, hence it does not hamper their functionality. 

2. It filters the variables when using the `data_manager.get_data()` method in the `KPI_Calculator`, so that now only the needed variables are retrieved. Each of these variables need to be reindexed, so retriving only the needed ones is a notable improvement.  

Note that with this new approach the KPI Calculator needs to reset every time the scenario is changed, so that the integration in the new scenario is not based on the results from the previous scenario.

### CODE REVIEW

**Comment by JavierArroyoBastida** [comment]
Unit tests are passing locally. For some reason Travis tests have not been triggered so far despite what has been discussed in https://github.com/ibpsa/project1-boptest/pull/263#issuecomment-743240878.

**Comment by dhblum** [comment]
Thanks @JavierArroyoBastida.  I likely won't be able to review this until next week.  And I think the tests are running now.  We need to form a queue with all other ibpsa repos, including modelica-ibpsa.  So sometimes, we'll have to wait until its our turn for them to start.

**Comment by JavierArroyoBastida** [comment]
Indeed, I see now they've already run. Thanks for clarifying @dhblum!

File: data/data_manager.py
Comment (dhblum) [inline data/data_manager.py:240]: Just a note that this may come in handy later too to make forecast data retrieval more efficient.

File: kpis/kpi_calculator.py
Comment (dhblum) [inline kpis/kpi_calculator.py:80]: Why the use of ``setattr`` instead of ``self.x = y``?

**Review by dhblum — CHANGES_REQUESTED** [review]
Thanks a lot @JavierArroyoBastida for this implementation.  In general it looks good and seems to do the job.  Just one question I had, see the inline comment.  Before merging, still need to:

- merge latest master and resolve conflicts
- pass pre-commit checks

**Comment by JavierArroyoBastida** [comment]
Thanks a lot for your review @dhblum! I've just merged master and resolved conflicts, waiting now for tests to pass.

File: kpis/kpi_calculator.py
Comment (JavierArroyoBastida) [reply kpis/kpi_calculator.py:80]: This is just to enable an attribute with a dynamic name that can be defined from a string using `format()`

Comment (dhblum) [reply kpis/kpi_calculator.py:80]: Ah ok I see now.  Ok sounds good, thanks.

**Comment by dhblum** [comment]
Thanks @JavierArroyoBastida for resolving the conflicts and merging latest master.  I ran pre-commit checks and there were some whitespace issues so committed a few additional changes.  Will merge after tests pass.

**Review by dhblum — APPROVED** [review]

**Pull request merged**
